### PR TITLE
Implement fallback hashing algorithm when crypto module is not available

### LINKED ISSUE #19100: Handle absent 'crypto' as per nodejs docs
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

TypeScript compiler relies on node's 'crypto' module unconditionally:

https://github.com/Microsoft/TypeScript/blob/6997e9b7316b081f8731d131139c7d7f1bba9092/src/compiler/sys.ts#L132

Whereas node's <a href="https://nodejs.org/api/crypto.html#crypto_determining_if_crypto_support_is_unavailable">documentation for crypto module</a> explicitly states:

<blockquote>
<h3>Determining if crypto support is unavailable</h3>
It is possible for Node.js to be built without including support for the crypto module. In such cases, calling require('crypto') will result in an error being thrown.

```javascript
let crypto;
try {
  crypto = require('crypto');
} catch (err) {
  console.log('crypto support is disabled!');
}
```
</blockquote>

That means on a reduced node environment the compiler crashes unnecessarily. By unnecessarily I mean the support for crypto is already optional in the compiler:

https://github.com/Microsoft/TypeScript/blob/6997e9b7316b081f8731d131139c7d7f1bba9092/src/compiler/sys.ts#L67

https://github.com/Microsoft/TypeScript/blob/d7269f1386bb1ad4351a6f6d5f749e89cb14c2b7/src/compiler/watch.ts#L635

https://github.com/Microsoft/TypeScript/blob/67a6a9477f90d61955f92fb33e7b11491dbd17a5/src/compiler/program.ts#L165

**Expected behavior:**

TSC to work without 'crypto'

**Actual behavior:**

TSC fails without 'crypto'

### PR DESCRIPTION
**Pull request by bttf**
Improves upon #19101; Fixes #19100

### CODE REVIEW

File: src/compiler/sys.ts [hunk 1]
```diff
@@ -494,9 +501,18 @@ namespace ts {
...
-                    return hash.digest("hex");
+                    let hash: any;
+                    if (_crypto) {
+                      hash = _crypto.createHash("md5");
+                      hash.update(data);
+                      hash.digest("hex");
+                    }
+                    else {
+                      // djb2 hashing algorithm: http://www.cse.yorku.ca/~oz/hash.html
+                      const chars = data.split("").map(str => str.charCodeAt(0));
```
Comment (sheetalkamat) [inline src/compiler/sys.ts]: Can you make this a exported function with ```/*@internal*/``` and also use it at https://github.com/Microsoft/TypeScript/blob/master/src/compiler/watch.ts#L645 in the ```system.createHash``` ```undefined``` case.

**Comment by mihailik** [comment]
Thanks for the addition @bttf , I've been lazy on this one.

Just looking at the original comments, one place concerns me: https://github.com/Microsoft/TypeScript/blob/0d9bc55033a3b5c8f68622803409476c7a6f82b4/src/compiler/sys.ts#L64

Why does it say 'should be cryptographically secure'?

I <a href="https://github.com/Microsoft/TypeScript/pull/19101#issuecomment-337033393">did raise</a> this question:

> @mhegazy just to check: collisions due to weak hash are only affecting performance, cannot cause incorrect behaviour, is that right?

reply:
> that is my expectation as well.

Looking at the usage in watch.ts the answer isn't that clear. @sheetalkamat you've put lots of work in there, can you clarify please?

How does the logic in `watch.ts` handle hash collisions?

Many thanks!

**Comment by sheetalkamat** [comment]
@mihailik The hash is created from .d.ts(so that we know shape of the module). If hash1 corresponds to text1 and hash2 corresponds to text2 then when shape of the module changes from text1 to text2, if hash1 = hash2 it would mean that we wont emit any other modules that depend on (import directly or indirectly).

**Comment by mihailik** [comment]
That backtracks on previous clarification. And it means the current released TSC is **guaranteed** to produce incorrect build results randomly (albeit rarely).

Also the risk with weak hash is higher.

Should the weak hash implmentation instead return the argument as result, as testing harness does that already: https://github.com/Microsoft/TypeScript/blob/7a4557331195e982abb7e1d493b42d185e301857/src/harness/harness.ts#L761

File: src/compiler/sys.ts [hunk 2]
```diff
@@ -16,6 +16,16 @@ namespace ts {
         }
     }
 
+    /**
+     * djb2 hashing algorithm
+     * http://www.cse.yorku.ca/~oz/hash.html
+     */
+    /* @internal */
+    export function generateDjb2Hash(data: string): string {
```
Comment (mhegazy) [inline src/compiler/sys.ts]: no need for this to be exported i would say. just make createHash always implemented in sys.

File: src/compiler/sys.ts [hunk 3]
```diff
@@ -494,9 +511,16 @@ namespace ts {
                     }
                 },
                 createHash(data) {
```
Comment (mhegazy) [inline src/compiler/sys.ts]: consider defining two functions instead:
```ts
function createMD5HashUsingNativeCrypto(data: string) {
    const hash = _crypto.createHash("md5");
    hash.update(data);
    return hash.digest("hex");
}
function createDjb2Hash(data) {
    const chars = data.split("").map(str => str.charCodeAt(0));
    return `${chars.reduce((prev, curr) => ((prev << 5) + prev) + curr, 5381)}`;
}
```
then in here make createHash either one or the other based on the existance of _crypto,e.g.:

```ts
{
...
createHash: _crypto ? createMD5HashUsingNativeCrypto : createDjb2Hash
...
}
```

File: src/compiler/watch.ts [hunk 4]
```diff
@@ -642,7 +642,7 @@ namespace ts {
         }
 
         function computeHash(data: string) {
-            return system.createHash ? system.createHash(data) : data;
+            return system.createHash ? system.createHash(data) : generateDjb2Hash(data);
```
Comment (mhegazy) [inline src/compiler/watch.ts]: just use `createHash` when available. you already made it use `generateDjb2Hash` in sys. watch is no triggered except on node anyways.

**Comment by mhegazy** [comment]
@sheetalkamat what about @mihailik suggestion of just returning the string.. it is definitely more accurate..

**Comment by sheetalkamat** [comment]
We can use same string but then it also means that huge string is stored in cache and would be there for each output for life time (till it changes to something else) (in tsc --w or compile on save from editor)

**Comment by mhegazy** [comment]
good point.

**Comment by mhegazy** [comment]
@bttf i think this change i ready to go. i am just in the middle of snapping for TS 2.7. i will merge this once that is done.

**Pull request merged**